src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs

```diff
@@ -219,45 +219,72 @@ private SafeProcessHandle GetProcessHandle()
+        /// Even with UseShellExecute option, we try first running fileName just in case the caller is giving executable which we should run
+        /// Then if we couldn't run it we'll try the shell tools to launch it(e.g. "open fileName")
+        /// </summary>
         /// <param name="startInfo">The start info with which to start the process.</param>
         private bool StartCore(ProcessStartInfo startInfo)
         {
             string filename;
             string[] argv;
 
+            if (Directory.Exists(startInfo.FileName))
+            {
+                throw new Win32Exception(SR.DirectoryNotValidAsInput);
+            }
+
             if (startInfo.UseShellExecute)
             {
                 if (startInfo.RedirectStandardInput || startInfo.RedirectStandardOutput || startInfo.RedirectStandardError)
                 {
                     throw new InvalidOperationException(SR.CantRedirectStreams);
                 }
+            }
 
-                const string ShellPath = "/bin/sh";
+            int childPid = -1, stdinFd = -1, stdoutFd = -1, stderrFd = -1, result = -1;
+            string[] envp = CreateEnvp(startInfo);
+            string cwd = !string.IsNullOrWhiteSpace(startInfo.WorkingDirectory) ? startInfo.WorkingDirectory : null;
 
-                filename = ShellPath;
-                argv = new string[3] { ShellPath, "-c", startInfo.FileName + " " + startInfo.Arguments};
-            }
-            else
+            filename = ResolvePath(startInfo.FileName); 
+            if (!string.IsNullOrEmpty(filename))
             {
-                filename = ResolvePath(startInfo.FileName);
                 argv = ParseArgv(startInfo);
+
+                // Invoke the shim fork/execve routine.  It will create pipes for all requested
+                // redirects, fork a child process, map the pipe ends onto the appropriate stdin/stdout/stderr
+                // descriptors, and execve to execute the requested process.  The shim implementation
+                // is used to fork/execve as executing managed code in a forked process is not safe (only
+                // the calling thread will transfer, thread IDs aren't stable across the fork, etc.)
+                result = Interop.Sys.ForkAndExecProcess(
+                    filename, argv, envp, cwd,
+                    startInfo.RedirectStandardInput, startInfo.RedirectStandardOutput, startInfo.RedirectStandardError,
+                    out childPid,
+                    out stdinFd, out stdoutFd, out stderrFd, shouldThrow: !startInfo.UseShellExecute);
             }
 
-            string[] envp = CreateEnvp(startInfo);
-            string cwd = !string.IsNullOrWhiteSpace(startInfo.WorkingDirectory) ? startInfo.WorkingDirectory : null;
+            if (result != 0)
+            {
+                if (!startInfo.UseShellExecute)
+                {
+                    // Could not find the file
+                    Debug.Assert(string.IsNullOrEmpty(filename));
```

Comment:
If filename was null/empty (and USE=false), then it would go into this block. so why assert? Debug.Assert should only be to catch mistakes by us (you). In this case it seems like it's for bad user input, and we should throw somewhere.

Reply:
This assert is against filename. 
- filename is retrieved from ResolvePath() on line 250. This happens when startInfo.FileName is a URL.
I added a comment on the summary of this method to explain this.